rr_supervisor/src/supervisor.cpp

```diff
@@ -1,16 +1,168 @@
 /** @file supervisor.cpp
  *  @brief Supervisor class implementation
- *  @author Waleed Ahmed(w29ahmed)
+ * 
+ *  *  Services provided:
+ *    /Supervisor/start_race
+ *    /Supervisor/count_lap
+ *
+ *  Topics Published:
+ *    /Supervisor/cmd
+ *    /Supervisor/no_movement
+ *    /Supervisor/enable_movement
+ *
+ *  Topics Subscribed:
+ *    /rr_vehicle/vel_cmd
+ *    /arduino/battery_state
+ *  @author Waleed Ahmed(w29ahmed) && Yuchi(Allan) Zhao
  *  @competition IARRC 2019
  */
 
 #include <supervisor.hpp>
 
+using namespace supervisor_topic;
+using namespace arduino_topic;
+using namespace rr_cmd_topics;
+
 /**
  * @brief Supervisor class constructor
  */
-Supervisor::Supervisor() {
+Supervisor::Supervisor() 
+{
+    raceStarted=false;
+    twist_msg_count_ = 0;
+
+    null_vector_.x = 0.0; null_vector_.y = 0.0; null_vector_.z = 0.0;
+    twist_msg_.linear = null_vector_;
+    twist_msg_.angular = null_vector_;
+
+      // Setup publishers for twist multiplexer
+    twist_pub_ = nh_.advertise<geometry_msgs::Twist>(cmd, 1);
+    null_lock_ = nh_.advertise<std_msgs::Bool>(no_movement, 1);
+
+    // Setup subscribers to monitor battery and speed
+    cmd_sub_ = nh_.subscribe(vel_cmd, 1, &Supervisor::TrackSpeed, this);
+    battery_sub_ = nh_.subscribe(battery_state, 1, &Supervisor::MonitorBattery, this);
+
+    // Setup service servers
+    start_race_service_ = nh_.advertiseService(start_race, &Supervisor::StartRace, this);
+    count_lap_service_  = nh_.advertiseService(count_lap, &Supervisor::CountLap, this);
+
+}
+
+Supervisor::~Supervisor() 
+{
+    raceStarted=false;
+    twist_msg_count_ = 0;
+    null_vector_.x = 0.0; null_vector_.y = 0.0; null_vector_.z = 0.0;
+    twist_msg_.linear = null_vector_;
+    twist_msg_.angular = null_vector_;
+}
+
+// other methods
+void Supervisor::IdleRobot()
+{
+    ROS_INFO(¨Robot Idle!¨);
+    //Set null_lock as true. Will NOT allow path planner and joy messages through
+    bool_msg_.data=true;
+    null_lock_.publish(bool_msg_);
+}
+
+// Callback methods
+bool Supervisor::StartRace(std_srvs::Empty::Request &req, std_srvs::Empty::Response &res)
+{
+    raceStarted=true;
+    //Set null_lock as false. Will allow path planner and joy messages through
+    bool_msg_.data=false;
+    null_lock_.publish(bool_msg_);
+
+    begin_time =clock();
+    ROS_INFO(¨Race started!¨);
+    return true;
 }
 
-Supervisor::~Supervisor() {
+bool Supervisor::CountLap(std_srvs::Trigger::Request &req, std_srvs::Trigger::Response &res)
+{
+    if(race_type==¨drag¨)
+    {
+        lap_count_+=1;
+        res.success=true;
+        ROS_INFO(¨Drag race complete! shutting down¨)
+    }
+    else if(race_type_==¨circuit¨)
```

Comment:
Circuit race for this year only has 1 lap, and that was the only race that had more than one 'lap' (aka passing an endline). While we could just scrap this lap counting functionality all together, I say we keep it in and make is so each race has to specify # of laps, so in future when/if they change a race to have more than 1 lap it should be a very simple change in the code!